### PR TITLE
Only read lazy references spec on first access

### LINKED ISSUE #1520: `LazyReferenceMapper` opens connections during deserialization / initialization
I haven't checked if this is true for all classes in fsspec but I believe it should be considered best practice that basically all objects exposed by `fsspec` are only performing IO requests when needed.

The implementation of `LazyReferenceMapper` currently performs a `cat_file` in its `__init__`, see https://github.com/fsspec/filesystem_spec/blob/a408121d41e9917e37010a4c1fbee4392aff0e61/fsspec/implementations/reference.py#L135-L137

This is particularly hurtful if that object is being (de-)serialized or even just initialized when another object is being deserialized.

For example, the following exception occurred in a dask context where this kind of object was submitted.

<details>

<summary>Long and a little boring traceback about a connection failure in pickle.loads</summary>

```python-traceback
aiohttp.client_exceptions.ServerTimeoutError: Connection timeout to host XXX/.zmetadata
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/coiled/env/lib/python3.11/site-packages/tornado/ioloop.py", line 738, in _run_callback
    ret = callback()
          ^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/tornado/ioloop.py", line 762, in _discard_future_result
    future.result()
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/worker.py", line 206, in wrapper
    return await method(self, *args, **kwargs)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/worker.py", line 1299, in handle_scheduler
    await self.handle_stream(comm)
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/core.py", line 1025, in handle_stream
    msgs = await comm.read()
           ^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/comm/tcp.py", line 247, in read
    msg = await from_frames(
          ^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/comm/utils.py", line 78, in from_frames
    res = _from_frames()
          ^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/comm/utils.py", line 61, in _from_frames
    return protocol.loads(
           ^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/protocol/core.py", line 175, in loads
    return msgpack.loads(
           ^^^^^^^^^^^^^^
  File "msgpack/_unpacker.pyx", line 194, in msgpack._cmsgpack.unpackb
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/protocol/core.py", line 172, in _decode_default
    return pickle.loads(sub_header["pickled-obj"], buffers=sub_frames)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/protocol/pickle.py", line 94, in loads
    return pickle.loads(x, buffers=buffers)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/spec.py", line 32, in make_instance
    return cls(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/spec.py", line 80, in __call__
    obj = super().__call__(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/implementations/cached.py", line 733, in __init__
    super().__init__(**kw)
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/implementations/cached.py", line 166, in __init__
    self.fs = fs if fs is not None else filesystem(target_protocol, **self.kwargs)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/registry.py", line 291, in filesystem
    return cls(**storage_options)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/spec.py", line 80, in __call__
    obj = super().__call__(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/implementations/reference.py", line 664, in __init__
    self.references = LazyReferenceMapper(
                      ^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/implementations/reference.py", line 135, in __init__
    self._items[".zmetadata"] = self.fs.cat_file(
                                ^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/asyn.py", line 118, in wrapper
    return sync(self.loop, func, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/asyn.py", line 103, in sync
    raise return_result
  File "/opt/coiled/env/lib/python3.11/site-packages/fsspec/asyn.py", line 56, in _runner
    result[0] = await coro
                ^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/s3fs/core.py", line 1102, in _cat_file
    return await _error_wrapper(_call_and_read, retries=self.retries)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/s3fs/core.py", line 140, in _error_wrapper
    raise err
  File "/opt/coiled/env/lib/python3.11/site-packages/s3fs/core.py", line 113, in _error_wrapper
    return await func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/s3fs/core.py", line 1089, in _call_and_read
    resp = await self._call_s3(
           ^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/s3fs/core.py", line 348, in _call_s3
    return await _error_wrapper(
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/s3fs/core.py", line 140, in _error_wrapper
    raise err
  File "/opt/coiled/env/lib/python3.11/site-packages/s3fs/core.py", line 113, in _error_wrapper
    return await func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/client.py", line 388, in _make_api_call
    http, parsed_response = await self._make_request(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/client.py", line 416, in _make_request
    return await self._endpoint.make_request(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/endpoint.py", line 100, in _send_request
    while await self._needs_retry(
          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/endpoint.py", line 262, in _needs_retry
    responses = await self._event_emitter.emit(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/hooks.py", line 66, in _emit
    response = await resolve_awaitable(handler(**kwargs))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/_helpers.py", line 15, in resolve_awaitable
    return await obj
           ^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/retryhandler.py", line 107, in _call
    if await resolve_awaitable(self._checker(**checker_kwargs)):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/_helpers.py", line 15, in resolve_awaitable
    return await obj
           ^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/retryhandler.py", line 126, in _call
    should_retry = await self._should_retry(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/retryhandler.py", line 165, in _should_retry
    return await resolve_awaitable(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/_helpers.py", line 15, in resolve_awaitable
    return await obj
           ^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/retryhandler.py", line 174, in _call
    checker(attempt_number, response, caught_exception)
  File "/opt/coiled/env/lib/python3.11/site-packages/botocore/retryhandler.py", line 247, in __call__
    return self._check_caught_exception(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/botocore/retryhandler.py", line 416, in _check_caught_exception
    raise caught_exception
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/endpoint.py", line 181, in _do_get_response
    http_response = await self._send(request)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/endpoint.py", line 285, in _send
    return await self.http_session.send(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/aiobotocore/httpsession.py", line 245, in send
    raise ConnectTimeoutError(endpoint_url=request.url, error=e)
botocore.exceptions.ConnectTimeoutError: Connect timeout on endpoint URL:
```

</details>

Getting these kinds of IO exceptions during (de-)serialization is quite uncommon and I doubt are handled very gracefully (at least they aren't in dask). Apart from the exceptions, this poses quite a significant performance problem if many of those objects are (de-)serialized

### PR DESCRIPTION
**Pull request by martindurant**
Fixes #1520 

@fjetter , would you also need a `__reduce__`, or is this enough?

### CODE REVIEW

File: fsspec/implementations/reference.py [hunk 1]
```diff
@@ -128,22 +128,32 @@ def __init__(
         """
         self.root = root
         self.chunk_sizes = {}
-        self._items = {}
+        self.out_root = out_root or self.root
+        self.cat_thresh = categorical_threshold
+        self.cache_size = cache_size
         self.dirs = None
+        self.url = self.root + "/{field}/refs.{record}.parq"
```
[the remark this answers is not in the document]
Comment (martindurant) [reply fsspec/implementations/reference.py:135]: Nope this was wrong. We don't want to format it here - you fooled me.

**Comment by rsignell** [comment]
Should I give this a try?

**Comment by martindurant** [comment]
@rsignell , go ahead, until I am told that it needs more

**Comment by fjetter** [comment]
Thanks for the quick reaction. IIUC a `__reduce__` is not required. This appears to be triggered by the deserialization of a `ReferenceFileSystem` which comes with a reduce that goes via `make_instance` and triggers the `LazyReferenceMapper.__init__` here https://github.com/fsspec/filesystem_spec/blob/303a0858dd9e3aea131b9e2274d22150d6d4f30e/fsspec/implementations/reference.py#L664

**Comment by martindurant** [comment]
OK got it - the reduce of the FS takes care of this, since it only cares about arguments that were used to create the instance.

**Pull request merged**